integration/spark/vendor/iceberg/src/main/java/io/openlineage/spark/agent/vendor/iceberg/metrics/IcebergMetricsReporterInjector.java

```diff
@@ -66,24 +67,38 @@ public boolean isDefinedAt(LogicalPlan plan) {
     return catalog.get().getClass().getCanonicalName().startsWith("org.apache.iceberg");
   }
 
-  private Optional<TableCatalog> getCatalog(LogicalPlan plan) {
+  /**
+   * @param plan
+   * @return
+   */
+  private Optional<CatalogPlugin> getCatalog(LogicalPlan plan) {
+    Optional<CatalogPlugin> catalog = getCatalogFromCaseClass(plan);
+    if (catalog.isPresent()) {
+      return catalog;
+    }
+
+    if (plan instanceof UnaryCommand) {
+      return getCatalogFromCaseClass(((UnaryCommand) plan).child());
+    } else if (plan instanceof BinaryCommand) {
+      return getCatalogFromCaseClass(((BinaryCommand) plan).left());
+    }
+
+    return Optional.empty();
+  }
+
+  /**
+   * Runs catalog method on LogicalPlan class. This is done through reflection as the case class
+   * with catalog property differs across different Spark versions. For example,
+   * `ResolvedDBObjectName` is available only for Spark 3.3
```

Comment:
In the cases where we're using reflection, we need to have complete explanation for where we're going into some different conditions, since we can't just ctrl+click on those using IDE. For example isn't good enough I believe, the explanation needs to be exhaustive, at least for those cases when we're not falling back to some default behavior.

I know it wasn't always done that way in the past but it's extremely confusing later, and there's no better moment to document that behavior than now.